eventhubs/data-plane/azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/CompleteLease.java

```diff
@@ -75,4 +77,18 @@ public long incrementEpoch() {
         this.epoch++;
         return this.epoch;
     }
+
+    @Override
+    public boolean equals(Object o) {
```

Comment:
@JonathanGiles  I found the we should not add equals() and hashCod() to CompleteLease, InMemoryLease, and AzureBlobLease. The reason is CompleteLease extends BaseLease, and InMemoryLease and AzureBlobLease extends CompleteLease. "There is no way to extends an instantiable class and add a value componet while preserving the equals contract" from Effective Java 3rd Edition Page 42.  It is a fundamental problem of equivalence relations in object-oriented languages.  I should not add equals() and hashCode() to CompleteLease, InMemoryLease and suppress them instead. What do you think?

Reply:
None of these classes implement equals / hashCode, do they? In that case, it seems reasonable to add them appropriately.

Reply:
None of these classes implement equals and hashCode initially. With only BaseLease has equals and hashCode. The CompleteLease, InMemoryLease and AzureBlobLease, have produced another SpotBugs error, "This class extends a class that defines an equals method and adds fields, but doesn't define an equals method itself. Thus, equality on instances of this class will ignore the identity of the subclass and the added fields. Be sure this is what is intended, and that you don't need to override the equals method. Even if you don't need to override the equals method, consider overriding it anyway to document the fact that the equals method for the subclass just return the result of invoking super.equals(o). "

Reply:
This seems like good advice from SpotBugs. I would suggest you consider if subclasses need to override the methods, or if they should just call super. You need to determine if there are additional properties on the subclasses that are important for testing equality.

Reply:
Since the existing design is using inheritance instead of composition. It will violate Symmetric rule of equals contract (a.equals(b)  == b.equals(a) )  such as if CompleteLease has included one additional properties than BaseLease,  CompleteLease.equals(BaseLease) return false, but BaseLease.equals(CompleteLease) return true.   So I believe it is better either just BaseLease's equals() method, or suppress it (since it doesn't help)

Reply:
I would call super.equals - I don't think suppressing is the best course of action here.